### DISCUSSION TITLE
logging doesn't work when using a c++ class header-only approach

### DISCUSSION THREAD

**Discussion by chmorgan** [zephyrproject-rtos/zephyr#79280] [General]
I've been using header-only C++ for a while now, it's convenient not having to split source and header apart. However this isn't compatible with the Zephyr logging system.

For example one cannot do something in example.h like:

```c++
class Example {
private:
    LOG_MODULE_REGISTER(Example, LOG_LEVEL_DBG);

public:
    void some_method()
    {
        LOG_INF("example log info");
    }
};
```

or even putting LOG_MODULE_REGISTER() outside of the class doesn't work either (as example.h may be included in several locations, so the LOG_MODULE_REGISTER() is duplicated several times).

I didn't see any examples that did this with the logging system, and its been a pain separating out .h/.cpp just for the logging system (I'm used to espressif's esp-idf where the logger module is a string passed to each log call and no LOG_MODULE_REGISTER() like thing is necessary. esp-idf logging works great even if used in header only c++ projects).

Thoughts? Maybe I'm missing something relative to advanced usage?

**Comment by dkalowsk** [comment]
I don't think this is a bug.  `LOG_MODULE_REGISTER` needs to be in a C/CPP file.  I think what you want is the `LOG_MODULE_DECLARE`.  Moving to a discussion.

**Reply by chmorgan** [reply]
@dkalowsk even if I did need a .cpp file for the LOG_MODULE_REGISTER(), could I use LOG_MODULE_DECLARE and put my log calls in the header file? I haven't seen an example of that in the repo.

**Reply by nordic-krch** [reply]
An example of logging from header using `LOG_MODULE_DECLARE` is here:
https://github.com/zephyrproject-rtos/zephyr/blob/0858399ffaf2efcab4afeb1647d80718debba032/samples/subsys/logging/logger/src/sample_module.h#L22

**Comment by oguzkaganozt** [comment]
I also noticed that logging mechanism seems unnecessarily complex and a bit unconventional. This approach makes debugging and maintaining the code harder. Additionally, I found that part of the documentation a bit hard to follow.

**Comment by nordic-krch** [comment]
@oguzkaganozt logging mechanism is complex but i don't agree that it's unnecessary. Logging is capable of building a logging message during compilation and not runtime and that makes it super fast compared to standard string formatting allowing non-intrusive logging from any context. There are configurations where logging strings are removed from binary (dictionary logging) and many more features that would not be possible without complex preprocessor operations.
